File: tools/dartdoc-compare.rb
require 'optparse'
require 'tmpdir'
require 'yaml'

options = {markdown_after: 'HEAD', sdk: false}

OptionParser.new do |opts|
  opts.banner = "Usage: dartdoc-compare.rb [OPTIONS] <dart-package>"

  opts.on("--dartdoc-dir=DARTDOC_DIR", "Directory of the dartdoc package") do |dartdoc_dir|
    options[:dartdoc_dir] = dartdoc_dir
  end

  opts.on("--before=BEFORE", "Markdown package 'before' ref") do |before|
    options[:markdown_before] = before
  end

  opts.on("--after=AFTER", "Markdown package 'after' ref (default: HEAD)") do |after|
    options[:markdown_after] = after
  end

  opts.on("--sdk", "Is the package the SDK?") do
    options[:sdk] = true
  end
end.parse!

raise OptionParser::MissingArgument.new("--dartdoc-dir") if options[:dartdoc_dir].nil?
raise OptionParser::MissingArgument.new("--before") if options[:markdown_before].nil?

class DartdocCompare
  def initialize(options)
    @dartdoc_dir = options[:dartdoc_dir]
    @markdown_before = options[:markdown_before]
    @markdown_after = options[:markdown_after]
    @dartdoc_bin = "#{@dartdoc_dir}/bin/dartdoc.dart"
    @dartdoc_pubspec_path = "#{@dartdoc_dir}/pubspec.yaml"
    @sdk = options[:sdk] ? '--sdk-docs --dart-sdk=%s' : ''
  end

  def compare(package)
    # Generate docs with Markdown "A".
    out_before = run_dartdoc(@markdown_before, package)

    # Generate docs with Markdown "B".
    out_after = run_dartdoc(@markdown_after, package)

    # Compare outputs
    diff = "diff -r -B #{out_before} #{out_after}"
    puts "Diff lines: #{`#{diff} |wc -l`.to_i}"
    puts diff
  end

  def run_dartdoc(markdown_ref, package)
    puts '=' * 80
    puts "Running dartdoc for #{markdown_ref}..."
    puts '=' * 80
    update_dartdoc_pubspec(markdown_ref)
    Dir.chdir(package)
    system('pub get')
    out = Dir.mktmpdir("dartdoc-compare-#{markdown_ref}__")
    sdk_options = @sdk % package
    cmd = "dart #{@dartdoc_bin} --output=#{out} #{sdk_options}"
    puts "Command: #{cmd}"
    system(cmd)
    puts ''

    out
  end

  def update_dartdoc_pubspec(markdown_ref)
    Dir.chdir(@dartdoc_dir)
    dartdoc_pubspec = YAML.load(File.read(@dartdoc_pubspec_path))

    dartdoc_pubspec['dependencies']['markdown'] = {
      'git' => {
        'url' => 'git://github.com/dart-lang/markdown.git',
        'ref' => markdown_ref
      }
    }

    File.write(@dartdoc_pubspec_path, YAML.dump(dartdoc_pubspec))
    system('pub get')
  end
end

package_to_document = ARGV.shift

comparer = DartdocCompare.new(options)
comparer.compare(package_to_document)

File: tool/dartdoc_compare.dart
// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'dart:convert' show jsonEncode, jsonDecode;
import 'dart:io' show Directory, File, Platform, Process, exitCode;

import 'package:args/args.dart' show ArgParser;
import 'package:path/path.dart' show absolute;
import 'package:yaml/yaml.dart' show loadYaml;

const _dartdocDir = 'dartdoc-dir';
const _markdownBefore = 'before';
const _markdownAfter = 'after';
const _sdk = 'sdk';
const _help = 'help';

void main(List<String> arguments) {
  final parser = ArgParser()
    ..addSeparator('Usage: dartdoc-compare.dart [OPTIONS] <dart-package>')
    ..addOption(_dartdocDir, help: 'Directory of the dartdoc package')
    ..addOption(_markdownBefore, help: "Markdown package 'before' ref")
    ..addOption(
      _markdownAfter,
      defaultsTo: 'HEAD',
      help: "Markdown package 'after' ref (or 'local')",
    )
    ..addFlag(
      _sdk,
      defaultsTo: false,
      negatable: false,
      help: 'Is the package the SDK?',
    )
    ..addFlag(_help, abbr: 'h', hide: true);

  final options = parser.parse(arguments);
  if (options[_help] as bool) {
    print(parser.usage);
    exitCode = 0;
    return;
  }
  if (options[_dartdocDir] == null || options[_markdownBefore] == null) {
    print(
        'Invalid arguments: Options --$_dartdocDir and --$_markdownBefore must be specified');
    print(parser.usage);
    exitCode = 1;
    return;
  }
  final comparer = DartdocCompare(
    options[_dartdocDir] as String,
    options[_markdownBefore] as String,
    options[_markdownAfter] as String,
    absolute(options[_dartdocDir] as String, 'bin/dartdoc.dart'),
    absolute(options[_dartdocDir] as String, 'pubspec.yaml'),
    options[_sdk] as bool,
  );

  String? path;
  if (comparer.sdk) {
    if (options.rest.isNotEmpty) {
      path = options.rest.single;
    }
  } else {
    path = options.rest.single;
  }

  if (comparer.compare(path)) {
    exitCode = 0;
  } else {
    exitCode = 1;
  }
}

class DartdocCompare {
  final String dartdocDir;
  final String markdownBefore;
  final String markdownAfter;
  final String dartdocBin;
  final String dartdocPubspecPath;
  final bool sdk;
  final String markdownPath = File(Platform.script.path).parent.parent.path;

  DartdocCompare(
    this.dartdocDir,
    this.markdownBefore,
    this.markdownAfter,
    this.dartdocBin,
    this.dartdocPubspecPath,
    this.sdk,
  );

  bool compare(String? package) {
    // Generate docs with Markdown "Before".
    final outBefore = _runDartdoc(markdownBefore, package);

    // Generate docs with Markdown "After".
    final outAfter = _runDartdoc(markdownAfter, package);

    // Compare outputs
    final diffOptions = ['-r', '-B', outBefore, outAfter];
    final result = Process.runSync('diff', diffOptions, runInShell: true);
    final nlines = '\n'.allMatches(result.stdout as String).length;
    print('Diff lines: $nlines');
    print('diff ${diffOptions.join(" ")}');
    return result.exitCode == 0;
  }

  String _runDartdoc(String markdownRef, String? path) {
    print('==========================================================');
    print('Running dartdoc for $markdownRef...');
    print('==========================================================');
    _doInPath(dartdocDir, () {
      final returnCode = _updateDartdocPubspec(markdownRef);
      if (returnCode != 0) {
        throw Exception("Could not update dartdoc's pubspec!");
      }
    });
    return _doInPath(path, () {
      if (!sdk) {
        _system('pub', ['upgrade']);
      }
      final out = Directory.systemTemp
          .createTempSync('dartdoc-compare-${markdownRef}__');
      final cmd = 'dart';
      final args = [dartdocBin, '--output=${out.path}'];

      if (sdk) {
        args.add('--sdk-docs');
      }

      print('Command: $cmd ${args.join(' ')}');
      final startTime = DateTime.now();
      _system(cmd, args);
      final endTime = DateTime.now();
      final duration = endTime.difference(startTime).inSeconds;
      print('dartdoc generation for $markdownRef took $duration seconds.');
      print('');

      return out.path;
    });
  }

  int _updateDartdocPubspec(String markdownRef) {
    var dartdocPubspec =
        loadYaml(File(dartdocPubspecPath).readAsStringSync()) as Map;
    // make modifiable copy
    dartdocPubspec = jsonDecode(jsonEncode(dartdocPubspec)) as Map;

    if (markdownRef == 'local') {
      dartdocPubspec['dependencies']['markdown'] = {
        'path': markdownPath,
      };
    } else {
      dartdocPubspec['dependencies']['markdown'] = {
        'git': {
          'url': 'git://github.com/dart-lang/markdown.git',
          'ref': markdownRef
        }
      };
    }

    File(dartdocPubspecPath).writeAsStringSync(jsonEncode(dartdocPubspec));
    return _system('pub', ['upgrade']);
  }
}

int _system(String cmd, List<String> args) {
  final result = Process.runSync(cmd, args);
  print(result.stdout);
  print(result.stderr);
  return result.exitCode;
}

T _doInPath<T>(String? path, T Function() f) {
  if (path == null) {
    return f();
  }

  final former = Directory.current.path;
  Directory.current = path;
  try {
    return f();
  } finally {
    Directory.current = former;
  }
}
